jib-core/src/main/java/com/google/cloud/tools/jib/event/progress/AllocationCompletionTracker.java

```diff
@@ -0,0 +1,162 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.jib.event.progress;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Keeps track of the progress for {@link Allocation}s as well as their order in which they appear.
+ *
+ * <p>This implementation is thread-safe.
+ */
+class AllocationCompletionTracker {
+
+  /**
+   * Holds the progress units completed along with a creation order. This is used as the value of
+   * the {@link #completionMap}.
+   */
+  private static class InsertionOrderUnits implements Comparable<InsertionOrderUnits> {
+
+    /** Monotonically-increasing source for {@link #index}. */
+    private static final AtomicInteger currentIndex = new AtomicInteger();
+
+    /** The creation order that monotonically increases. */
+    private final int index = currentIndex.getAndIncrement();
+
+    /**
+     * Progress units completed. This can be shared across multiple threads and should be updated
+     * atomically.
+     */
+    private final AtomicLong units = new AtomicLong();
+
+    /** The {@link Allocation} this is for. */
+    private final Allocation allocation;
+
+    private InsertionOrderUnits(Allocation allocation) {
+      this.allocation = allocation;
+    }
+
+    @Override
+    public int compareTo(InsertionOrderUnits otherInsertionOrderUnits) {
+      return index - otherInsertionOrderUnits.index;
+    }
+  }
+
+  /**
+   * Maps from {@link Allocation} to the number of progress units completed in that {@link
+   * Allocation}, while keeping track of the insertion order of the {@link Allocation}s with {@link
+   * InsertionOrderUnits}.
+   */
+  private final ConcurrentHashMap<Allocation, InsertionOrderUnits> completionMap =
+      new ConcurrentHashMap<>();
+
+  /**
+   * Updates the progress for {@link Allocation} atomically relative to the {@code allocation}.
+   *
+   * <p>For any {@link Allocation}, this method <em>must</em> have been called on all of its parents
+   * beforehand.
+   *
+   * @param allocation the {@link Allocation} to update progress for
+   * @param units the units of progress
+   * @return {@code true} if the map was updated; {@code false} if {@code allocation} was already
```

Comment:
I think it's easier to understand to say something along the line "it may return true even if the map was not updated if called concurrently, but never false if the map was updated."